x/emissions/module/rewards_v2.go

```diff
@@ -0,0 +1,430 @@
+	totalStake := 0.0
+	for _, stake := range reputersAdjustedStakes {
+		totalStake += stake
+	}
+
+	// Ensure every loss array is non-empty and calculate geometric mean
+	stakeWeightedLoss := make([]float64, len(reputersReportedLosses[0]))
+	for j := 0; j < len(reputersReportedLosses[0]); j++ {
+		logSum := 0.0
+		for i, losses := range reputersReportedLosses {
+			logSum += (math.Log10(losses[j]) * reputersAdjustedStakes[i]) / totalStake
+		}
+		stakeWeightedLoss[j] = logSum
+	}
+
+	return stakeWeightedLoss, nil
+}
+
+// GetConsensusScore calculates the proximity to consensus score for a reputer.
+// T_im
+func GetConsensusScore(reputerLosses, consensusLosses []float64) (float64, error) {
+	fTolerance := 0.01
+	if len(reputerLosses) != len(consensusLosses) {
+		return 0, types.ErrInvalidSliceLength
+	}
+
+	var sumLogConsensusSquared float64
+	for _, cLoss := range consensusLosses {
+		sumLogConsensusSquared += math.Pow(cLoss, 2) //!
+	}
+	consensusNorm := math.Sqrt(sumLogConsensusSquared)
+
+	var distanceSquared float64
+	for i, rLoss := range reputerLosses {
+		distanceSquared += math.Pow(math.Log10(rLoss/consensusLosses[i]), 2)
+	}
+	distance := math.Sqrt(distanceSquared)
+
+	score := 1 / (distance/consensusNorm + fTolerance)
+	return score, nil
+}
+
+// GetAllConsensusScores calculates the proximity to consensus score for all reputers.
+// calculates:
+// T_i - stake weighted total consensus
+// returns:
+// T_im - reputer score (proximity to consensus)
+func GetAllConsensusScores(allLosses [][]float64, stakes []float64, allListeningCoefficients []float64, numReputers int) ([]float64, error) {
+	// Get adjusted stakes
+	var adjustedStakes []float64
+	for i, reputerStake := range stakes {
+		adjustedStake, err := GetAdjustedStake(reputerStake, stakes, allListeningCoefficients[i], allListeningCoefficients, float64(numReputers))
+		if err != nil {
+			return nil, err
+		}
+		adjustedStakes = append(adjustedStakes, adjustedStake)
+	}
+
+	// Get consensus loss vector
+	consensus, err := GetStakeWeightedLossMatrix(adjustedStakes, allLosses)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get reputers scores
+	scores := make([]float64, numReputers)
+	for i := 0; i < numReputers; i++ {
+		losses := allLosses[i]
+		scores[i], err = GetConsensusScore(losses, consensus)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return scores, nil
+}
+
+// GetAllReputersOutput calculates the final scores and adjusted listening coefficients for all reputers.
+// This function iteratively adjusts the listening coefficients based on a gradient descent method to minimize
+// the difference between each reputer's losses and the consensus losses, taking into account each reputer's stake.
+// returns:
+// T_im - reputer score (proximity to consensus)
+// a_im - listening coefficients
+func GetAllReputersOutput(allLosses [][]float64, stakes []float64, initialCoefficients []float64, numReputers int, enableListening bool) ([]float64, []float64, error) {
```

Comment:
plz make a follow up ticket for after you stitch things tg to test for "bounded iterations" (as I said in other comment) but also to simultaneously clean and optimize this function. "coeffs, cofficients, coeffs2..." plus all the copies and repeated compute (e.g. you should be able to call GetAllConsensusScores once not twice; you probs also have more `copy` calls than you need). you implemented the code in the simulation faithfully but perhaps too much so -- pandas in the simulation lets you get away with more compute than an on-chain environment, so we need to care about optimizations here more than the simulation authors had to

Reply:
ticket created!